### ISSUE TITLE
W3C response body with falsy value is interpreted incorrectly

### ISSUE THREAD

**Issue by sherlock1982** [webdriverio/webdriverio#2394]
## The problem

There are requests with falsy responses like this:

    {
        "value": null
    }

Or like this:

    {
        "value": false
    }

Actual behavior:

Will be interpreted as unknown error because there's no status field. 

Expected behavior:

Return value because it's correct answer

## Details

In 4.8.0 it was broken but still not fixed in current master.
What finally worked for me in RequestHandler.js:

```
/**
 * Resolve with a healthy response
 */
if (!err && body && (body.status === 0 || (body.value === null && !body.status) || (body.value !== undefined && !body.status && !(body.value.error || body.value.stackTrace)))) {
    return resolve({body, response})
}

```

Can be reproduced with Geckodriver commands: /enabled, /execute/sync when a command returns false or undefined.

**Comment by sherlock1982** [comment]
Another way to reproduce is to run a /execute which will returns falsy. false, 0, undefined. 
Maybe instead of body.value we should check if the value is present in the body. And if it's present than it's the result. 

Not really sure about body.value.error and body.value.stackTrace.

**Comment by christian-bromann** [comment]
Not sure what you mean but I don't see any problems:
```js
> browser.execute(() => false)                                                                                         │2017-11-06 00:47:00.284 plugin-container[87689:32822845] *** CFMessagePort: bootstrap_register(): failed 1100 (0x44c)
{ state: 'success',                                                                                                    │'Permission denied', port = 0xaa3b, name = 'com.apple.tsm.portname'
  sessionId: '103e313c-4a11-41ad-a52f-366fc7404c03',                                                                   │See /usr/include/servers/bootstrap_defs.h for the error codes.
  hCode: 1807659609,                                                                                                   │1509925620592   Marionette      INFO    Listening on port 63865
  value: false,                                                                                                        │1509925620648   Marionette      DEBUG   Register listener.js for window 2147483649
  class: 'org.openqa.selenium.remote.Response',                                                                        │00:47:00.659 INFO - Detected dialect: W3C
  _status: 0 }                                                                                                         │00:47:00.660 INFO - Done: [new session: Capabilities [{rotatable=true, locationContextEnabled=true, loggingPrefs=org.o
> browser.execute(() => null)                                                                                          │penqa.selenium.logging.LoggingPreferences@4b073de6, browserName=firefox, javascriptEnabled=true, handlesAlerts=true, r
{ state: 'success',                                                                                                    │equestOrigins={name=webdriverio, version=4.8.0, url=http://webdriver.io}}]]
  sessionId: '103e313c-4a11-41ad-a52f-366fc7404c03',                                                                   │00:47:09.682 INFO - Executing: [find elements: By.cssSelector: foobar])
  hCode: 1136321649,                                                                                                   │00:47:09.692 INFO - Done: [find elements: By.cssSelector: foobar]
  value: null,                                                                                                         │00:47:39.141 INFO - Executing: [execute script: return (() => false).apply(null, arguments), []])
  class: 'org.openqa.selenium.remote.Response',                                                                        │00:47:39.158 INFO - Done: [execute script: return (() => false).apply(null, arguments), []]
  _status: 0 }                                                                                                         │00:47:43.674 INFO - Executing: [execute script: return (() => null).apply(null, arguments), []])
> browser.execute(() => ({ value: false }))                                                                            │lib/scripts/getHTML.js -> build/lib/scripts/getHTML.js
{ state: 'success',                                                                                                    │lib/scripts/getViewportSize.js -> build/lib/scripts/getViewportSize.js
  sessionId: '103e313c-4a11-41ad-a52f-366fc7404c03',                                                                   │lib/scripts/isWithinViewport.js -> build/lib/scripts/isWithinViewport.js
  hCode: 673690713,                                                                                                    │lib/scripts/newWindow.js -> build/lib/scripts/newWindow.js
  value: { value: false },                                                                                             │lib/scripts/scroll.js -> build/lib/scripts/scroll.js
  class: 'org.openqa.selenium.remote.Response',                                                                        │lib/utils/BaseReporter.js -> build/lib/utils/BaseReporter.js
  _status: 0 }                                                                                                         │lib/utils/ConfigParser.js -> build/lib/utils/ConfigParser.js
> browser.execute(() => ({ value: null }))                                                                             │lib/utils/ErrorHandler.js -> build/lib/utils/ErrorHandler.js
{ state: 'success',                                                                                                    │lib/utils/Logger.js -> build/lib/utils/Logger.js
  sessionId: '103e313c-4a11-41ad-a52f-366fc7404c03',                                                                   │lib/utils/ReporterStats.js -> build/lib/utils/ReporterStats.js
  hCode: 1116273200,                                                                                                   │lib/utils/RequestHandler.js -> build/lib/utils/RequestHandler.js
  value: { value: null },                                                                                              │lib/utils/Timer.js -> build/lib/utils/Timer.js
  class: 'org.openqa.selenium.remote.Response',                                                                        │lib/webdriverio.js -> build/lib/webdriverio.js
  _status: 0 }
```

**Comment by sherlock1982** [comment]
Yes because you have a status field. But status field is not a must have according to W3C. I use Geckodriver which returns false only. Here's the output with wdio runner:

```
COMMAND: GET http://selenium.3cx.test:4444/wd/hub/session/2f403804-755a-4b00-ab5a-2e1c86b7620b/element/39b11f68-3128-4997-bfcd-40385f2e1a14/displayed - {}
RESULT: {"value":false}

```

Now imagine what's gonna happen with this code in RequestHandler:

              /**
                 * Resolve with a healthy response
                 */
                if (!err && body && (body.status === 0 || (body.value && !body.status && !(body.value.error || body.value.stackTrace)))) {
                    return resolve({body, response})
                }

There's no status field and body.value===false. Which means that instead of calling resolve() it returns error.

**Comment by christian-bromann** [comment]
Are you running the latest Geckodriver version?

**Comment by sherlock1982** [comment]
0.19.0 but the issue is not with the driver. Driver answer is correct. It's just falsy. I attached proposed fix in the very beginning.
Sorry, sometimes I start explanation from the middle.

**Comment by sherlock1982** [comment]
Please note that same error can be reproduced with:

    browser.waitForEnabled();

In case element is disabled. This returns: { "value": false }

**Comment by sherlock1982** [comment]
Fixed in 4.9.9
